File: app/src/main/java/app/trigger/bluetooth/BluetoothRequestHandler.java
package app.trigger.bluetooth;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;

import java.io.IOException;
import java.util.Set;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.UUID;

import app.trigger.BluetoothDoorSetup;
import app.trigger.BluetoothTools;
import app.trigger.MainActivity.Action;
import app.trigger.DoorReply.ReplyCode;
import app.trigger.OnTaskCompleted;
import app.trigger.Log;


public class BluetoothRequestHandler extends Thread {
    private final OnTaskCompleted listener;
    private final BluetoothDoorSetup setup;
    private final Action action;
    private BluetoothSocket socket;

    public BluetoothRequestHandler(OnTaskCompleted listener, BluetoothDoorSetup setup, Action action) {
        this.listener = listener;
        this.setup = setup;
        this.action = action;
    }

    public void run() {
        if (setup.getId() < 0) {
            this.listener.onTaskResult(setup.getId(), ReplyCode.LOCAL_ERROR, "Internal Error");
            return;
        }

        BluetoothAdapter adapter = BluetoothAdapter.getDefaultAdapter();

        if (adapter == null) {
            this.listener.onTaskResult(setup.getId(), ReplyCode.DISABLED, "Device does not support Bluetooth");
            return;
        }

        if (!adapter.isEnabled()) {
            // request to enable
            this.listener.onTaskResult(setup.getId(), ReplyCode.DISABLED, "Bluetooth is disabled.");
            return;
        }

        String request = "";
        String response = "";

        switch (action) {
            case open_door:
                request = setup.open_query;
                break;
            case ring_door:
                request = setup.ring_query;
                break;
            case close_door:
                request = setup.close_query;
                break;
            case fetch_state:
                request = setup.status_query;
                break;
        }

        if (request.isEmpty()) {
            this.listener.onTaskResult(setup.getId(), ReplyCode.LOCAL_ERROR, "");
            return;
        }

        try {
            Set<BluetoothDevice> pairedDevices = adapter.getBondedDevices();

            String address = "";
            for (BluetoothDevice device : pairedDevices) {
                if ((device.getName() != null && device.getName().equals(setup.device_name))
					|| device.getAddress().equals(setup.device_name.toUpperCase())) {
                    address = device.getAddress();
                }
            }

            if (address.isEmpty()) {
                this.listener.onTaskResult(setup.getId(), ReplyCode.LOCAL_ERROR, "Device not paired yet.");
                return;
            }

            BluetoothDevice device = adapter.getRemoteDevice(address);

            if (setup.service_uuid.isEmpty()) {
                socket = BluetoothTools.createRfcommSocket(device);
            } else {
                UUID uuid = UUID.fromString(setup.service_uuid);
                socket = device.createRfcommSocketToServiceRecord(uuid);
            }

            socket.connect();

            // Get the BluetoothSocket input and output streams
            InputStream tmpIn = socket.getInputStream();
            OutputStream tmpOut = socket.getOutputStream();

            tmpOut.write(request.getBytes());
            tmpOut.flush();

            try {
                byte[] buffer = new byte[512];
                int bytes = tmpIn.read(buffer);
                response = new String(buffer, 0, bytes);
            } catch (IOException ioe) {
                this.listener.onTaskResult(setup.getId(), ReplyCode.REMOTE_ERROR, "Cannot reach remote device.");
                return;
            }

            socket.close();

            this.listener.onTaskResult(setup.getId(), ReplyCode.SUCCESS, response);
        } catch (Exception e) {
            this.listener.onTaskResult(setup.getId(), ReplyCode.LOCAL_ERROR, e.toString());
        }
    }
}

File: app/src/main/kotlin/app/trigger/bluetooth/BluetoothRequestHandler.kt
package app.trigger.bluetooth

import app.trigger.BluetoothTools.createRfcommSocket
import android.bluetooth.BluetoothSocket
import app.trigger.DoorReply.ReplyCode
import android.bluetooth.BluetoothAdapter
import app.trigger.*
import java.io.IOException
import java.lang.Exception
import java.util.*


class BluetoothRequestHandler(private val listener: OnTaskCompleted, private val setup: BluetoothDoorSetup, private val action: MainActivity.Action) : Thread() {
    private var socket: BluetoothSocket? = null
    override fun run() {
        if (setup.id < 0) {
            listener.onTaskResult(setup.id, ReplyCode.LOCAL_ERROR, "Internal Error")
            return
        }

        val adapter = BluetoothAdapter.getDefaultAdapter()
        if (adapter == null) {
            listener.onTaskResult(setup.id, ReplyCode.DISABLED, "Device does not support Bluetooth")
            return
        }

        if (!adapter.isEnabled) {
            // request to enable
            listener.onTaskResult(setup.id, ReplyCode.DISABLED, "Bluetooth is disabled.")
            return
        }

        val request = when (action) {
            MainActivity.Action.OPEN_DOOR -> setup.open_query
            MainActivity.Action.RING_DOOR -> setup.ring_query
            MainActivity.Action.CLOSE_DOOR -> setup.close_query
            MainActivity.Action.FETCH_STATE -> setup.status_query
        }
        if (request.isEmpty()) {
            listener.onTaskResult(setup.id, ReplyCode.LOCAL_ERROR, "")
            return
        }
        try {
            val pairedDevices = adapter.bondedDevices
            var address = ""
            for (device in pairedDevices) {
                if (device.name != null && device.name == setup.device_name
                        || device.address == setup.device_name.toUpperCase()) {
                    address = device.address
                }
            }
            if (address.isEmpty()) {
                listener.onTaskResult(setup.id, ReplyCode.LOCAL_ERROR, "Device not paired yet.")
                return
            }
            val device = adapter.getRemoteDevice(address)
            socket = if (setup.service_uuid.isEmpty()) {
                createRfcommSocket(device)
            } else {
                val uuid = UUID.fromString(setup.service_uuid)
                device.createRfcommSocketToServiceRecord(uuid)
            }
            socket!!.connect()

            // Get the BluetoothSocket input and output streams
            val tmpIn = socket!!.inputStream
            val tmpOut = socket!!.outputStream
            tmpOut.write(request.toByteArray())
            tmpOut.flush()
            val response = try {
                val buffer = ByteArray(512)
                val bytes = tmpIn.read(buffer)
                String(buffer, 0, bytes)
            } catch (ioe: IOException) {
                listener.onTaskResult(setup.id, ReplyCode.REMOTE_ERROR, "Cannot reach remote device.")
                return
            }
            socket!!.close()
            listener.onTaskResult(setup.id, ReplyCode.SUCCESS, response)
        } catch (e: Exception) {
            listener.onTaskResult(setup.id, ReplyCode.LOCAL_ERROR, e.toString())
        }
    }
}